src/internal/task/interruptqueue.go

```diff
@@ -0,0 +1,66 @@
+package task
+
+import "runtime/volatile"
+
+// InterruptQueue is a specialized version of Queue, designed for working with interrupts.
+// It can be safely pushed to from an interrupt (assuming that a memory reference to the task remains elsewhere), and popped outside of an interrupt.
+// It cannot be pushed to outside of an interrupt or popped inside an interrupt.
+type InterruptQueue struct {
+	// This implementation uses a double-buffer of queues.
+	// bufSelect contains the index of the queue currently available for pop operations.
+	// The opposite queue is available for push operations.
+	bufSelect volatile.Register8
+	queues    [2]Queue
+}
+
+// Push a task onto the queue.
+// This can only be safely called from inside an interrupt.
+func (q *InterruptQueue) Push(t *Task) {
+	// Avoid nesting interrupts inside here.
+	var nest nonest
+	nest.Lock()
+	defer nest.Unlock()
+
+	// Push to inactive queue.
+	q.queues[1-q.bufSelect.Get()].Push(t)
+}
+
+// Check if the queue is empty.
+// This will return false if any tasks were pushed strictly before this call.
+// If any pushes occur during the call, the queue may or may not be marked as empty.
```

Comment:
> the queue may or may not be marked as empty.

Could be me, but 'marked' sounds like it changes the queue, while it doesn't (at least, not in a relevant way). It only returns the current state of the queue.